`promtool test rules` fails when /tmp is not writable

### What did you do?

Tried to run `promtool` from the prometheus docker and it fails while it works for the same files

```
$ promtool test rules tests/unit/promtool/test_unit_available.yaml                                                                                                                         Unit Testing:  tests/unit/promtool/test_unit_available.yaml
  SUCCESS

$ docker run -v $PWD:/tmp --entrypoint "/bin/promtool" prom/prometheus:latest  test rules /tmp/tests/unit/promtool/test_unit_available.yaml                                              
Unit Testing:  /tmp/tests/unit/promtool/test_unit_available.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x19f3f48]

goroutine 1 [running]:
github.com/prometheus/prometheus/promql.(*LazyLoader).Errorf(0xc00067a000?, {0x34cdded?, 0xc000111c20?}, {0xc0006d7e10?, 0x2fcd000?, 0x56bb9a0?})
	<autogenerated>:1 +0x28
github.com/stretchr/testify/assert.Fail({0x7fcef76a0ba0, 0xc000111c20}, {0xc0000544b0, 0x4f}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:333 +0x350
github.com/stretchr/testify/assert.NoError({0x7fcef76a0ba0, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:1495 +0xfc
github.com/stretchr/testify/require.NoError({0x7fcef76a0b58, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/require/require.go:1357 +0x8f
github.com/prometheus/prometheus/util/teststorage.New({0x3bc6af8?, 0xc000111c20})
	/app/util/teststorage/storage.go:34 +0xc5
github.com/prometheus/prometheus/promql.(*LazyLoader).clear(0xc000111c20)
	/app/promql/test.go:772 +0xd6
github.com/prometheus/prometheus/promql.NewLazyLoader({0x0?, 0x0}, {0xc000492070, 0x68}, {0x60?, 0xac?})
	/app/promql/test.go:736 +0xa9
main.(*testGroup).test(0xc000a40fa0, 0xdf8475800, 0xc0005930e0?, {0xe0?, 0x8d?}, 0x0, {0xc0006d7d10, 0x1, 0x1})
	/app/cmd/promtool/unittest.go:180 +0xa5
main.ruleUnitTest({0x7ffd516b4f55, 0x31}, {0xa?, 0x0?}, 0xc000a41188?, 0x4c?)
	/app/cmd/promtool/unittest.go:114 +0x58c
main.RulesUnitTest({0x0?, 0x0?}, {0x0?, 0x3ba9140?, 0x1?}, 0x0?, {0xc0006d76e0?, 0x1, 0x0?})
	/app/cmd/promtool/unittest.go:54 +0x131
main.main()
	/app/cmd/promtool/main.go:373 +0xb278
```

### What did you expect to see?

Same output as when I download promtool binary e.g.
```
Unit Testing:  tests/unit/promtool/test_unit_available.yaml
  SUCCESS
```

### What did you see instead? Under which circumstances?

```
Unit Testing:  /tmp/tests/unit/promtool/test_unit_available.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x19f3f48]

goroutine 1 [running]:
github.com/prometheus/prometheus/promql.(*LazyLoader).Errorf(0xc00067a000?, {0x34cdded?, 0xc000111c20?}, {0xc0006d7e10?, 0x2fcd000?, 0x56bb9a0?})
	<autogenerated>:1 +0x28
github.com/stretchr/testify/assert.Fail({0x7fcef76a0ba0, 0xc000111c20}, {0xc0000544b0, 0x4f}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:333 +0x350
github.com/stretchr/testify/assert.NoError({0x7fcef76a0ba0, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:1495 +0xfc
github.com/stretchr/testify/require.NoError({0x7fcef76a0b58, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/require/require.go:1357 +0x8f
github.com/prometheus/prometheus/util/teststorage.New({0x3bc6af8?, 0xc000111c20})
	/app/util/teststorage/storage.go:34 +0xc5
github.com/prometheus/prometheus/promql.(*LazyLoader).clear(0xc000111c20)
	/app/promql/test.go:772 +0xd6
github.com/prometheus/prometheus/promql.NewLazyLoader({0x0?, 0x0}, {0xc000492070, 0x68}, {0x60?, 0xac?})
	/app/promql/test.go:736 +0xa9
main.(*testGroup).test(0xc000a40fa0, 0xdf8475800, 0xc0005930e0?, {0xe0?, 0x8d?}, 0x0, {0xc0006d7d10, 0x1, 0x1})
	/app/cmd/promtool/unittest.go:180 +0xa5
main.ruleUnitTest({0x7ffd516b4f55, 0x31}, {0xa?, 0x0?}, 0xc000a41188?, 0x4c?)
	/app/cmd/promtool/unittest.go:114 +0x58c
main.RulesUnitTest({0x0?, 0x0?}, {0x0?, 0x3ba9140?, 0x1?}, 0x0?, {0xc0006d76e0?, 0x1, 0x0?})
	/app/cmd/promtool/unittest.go:54 +0x131
main.main()
	/app/cmd/promtool/main.go:373 +0xb278
```

Other environment information:

```
$ docker version                                                                                                                            
Client:
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.8
 Git commit:        ced0996
 Built:             Sat Oct  7 00:14:00 2023
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          24.0.5
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.8
  Git commit:       a61e2b4
  Built:            Sat Oct  7 00:14:30 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### System information

Linux 6.5.0-21-generic x86_64

### Prometheus version

```text
2.50.0 but this happens with 2.20.0 image,2.30.0, 2.40.0
```


### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

--- a/cmd/promtool/unittest.go
+++ b/cmd/promtool/unittest.go
@@ -175,13 +175,18 @@ type testGroup struct {
 }
 
 // test performs the unit tests.
-func (tg *testGroup) test(evalInterval time.Duration, groupOrderMap map[string]int, queryOpts promql.LazyLoaderOpts, diffFlag bool, ruleFiles ...string) []error {
+func (tg *testGroup) test(evalInterval time.Duration, groupOrderMap map[string]int, queryOpts promql.LazyLoaderOpts, diffFlag bool, ruleFiles ...string) (outErr []error) {
 	// Setup testing suite.
-	suite, err := promql.NewLazyLoader(nil, tg.seriesLoadingString(), queryOpts)
+	suite, err := promql.NewLazyLoader(tg.seriesLoadingString(), queryOpts)
 	if err != nil {
 		return []error{err}
 	}
-	defer suite.Close()
+	defer func() {
+		err := suite.Close()
+		if err != nil {
+			outErr = append(outErr, err)
+		}
+	}()
 	suite.SubqueryInterval = evalInterval
 
 	// Load the rule files.

--- a/promql/test.go
+++ b/promql/test.go
@@ -704,8 +704,6 @@ func parseNumber(s string) (float64, error) {
 // LazyLoader lazily loads samples into storage.
 // This is specifically implemented for unit testing of rules.
 type LazyLoader struct {
-	testutil.T
-
 	loadCmd *loadCmd
 
 	storage          storage.Storage
@@ -727,13 +725,15 @@ type LazyLoaderOpts struct {
 }
 
 // NewLazyLoader returns an initialized empty LazyLoader.
-func NewLazyLoader(t testutil.T, input string, opts LazyLoaderOpts) (*LazyLoader, error) {
+func NewLazyLoader(input string, opts LazyLoaderOpts) (*LazyLoader, error) {
 	ll := &LazyLoader{
-		T:    t,
 		opts: opts,
 	}
 	err := ll.parse(input)
-	ll.clear()
+	if err != nil {
+		return nil, err
+	}
+	err = ll.clear()
 	return ll, err
 }
 
@@ -761,15 +761,20 @@ func (ll *LazyLoader) parse(input string) error {
 }
 
 // clear the current test storage of all inserted samples.
-func (ll *LazyLoader) clear() {
+func (ll *LazyLoader) clear() error {
 	if ll.storage != nil {
-		err := ll.storage.Close()
-		require.NoError(ll.T, err, "Unexpected error while closing test storage.")
+		if err := ll.storage.Close(); err != nil {
+			return fmt.Errorf("closing test storage: %w", err)
+		}
 	}
 	if ll.cancelCtx != nil {
 		ll.cancelCtx()
 	}
-	ll.storage = teststorage.New(ll)
+	var err error
+	ll.storage, err = teststorage.NewWithError()
+	if err != nil {
+		return err
+	}
 
 	opts := EngineOpts{
 		Logger:                   nil,
@@ -783,6 +788,7 @@ func (ll *LazyLoader) clear() {
 
 	ll.queryEngine = NewEngine(opts)
 	ll.context, ll.cancelCtx = context.WithCancel(context.Background())
+	return nil
 }
 
 // appendTill appends the defined time series to the storage till the given timestamp (in milliseconds).
@@ -836,8 +842,7 @@ func (ll *LazyLoader) Storage() storage.Storage {
 }
 
 // Close closes resources associated with the LazyLoader.
-func (ll *LazyLoader) Close() {
+func (ll *LazyLoader) Close() error {
 	ll.cancelCtx()
-	err := ll.storage.Close()
-	require.NoError(ll.T, err, "Unexpected error while closing test storage.")
+	return ll.storage.Close()
 }

--- a/promql/test_test.go
+++ b/promql/test_test.go
@@ -110,7 +110,7 @@ func TestLazyLoader_WithSamplesTill(t *testing.T) {
 	}
 
 	for _, c := range cases {
-		suite, err := NewLazyLoader(t, c.loadString, LazyLoaderOpts{})
+		suite, err := NewLazyLoader(c.loadString, LazyLoaderOpts{})
 		require.NoError(t, err)
 		defer suite.Close()
 

--- a/util/teststorage/storage.go
+++ b/util/teststorage/storage.go
@@ -14,6 +14,7 @@
 package teststorage
 
 import (
+	"fmt"
 	"os"
 	"time"
 
@@ -30,8 +31,18 @@ import (
 // New returns a new TestStorage for testing purposes
 // that removes all associated files on closing.
 func New(t testutil.T) *TestStorage {
+	stor, err := NewWithError()
+	require.NoError(t, err)
+	return stor
+}
+
+// NewWithError returns a new TestStorage for user facing tests, which reports
+// errors directly.
+func NewWithError() (*TestStorage, error) {
 	dir, err := os.MkdirTemp("", "test_storage")
-	require.NoError(t, err, "unexpected error while opening test directory")
+	if err != nil {
+		return nil, fmt.Errorf("opening test directory: %w", err)
+	}
 
 	// Tests just load data for a series sequentially. Thus we
 	// need a long appendable window.
@@ -41,13 +52,17 @@ func New(t testutil.T) *TestStorage {
 	opts.RetentionDuration = 0
 	opts.EnableNativeHistograms = true
 	db, err := tsdb.Open(dir, nil, nil, opts, tsdb.NewDBStats())
-	require.NoError(t, err, "unexpected error while opening test storage")
+	if err != nil {
+		return nil, fmt.Errorf("opening test storage: %w", err)
+	}
 	reg := prometheus.NewRegistry()
 	eMetrics := tsdb.NewExemplarMetrics(reg)
 
 	es, err := tsdb.NewCircularExemplarStorage(10, eMetrics)
-	require.NoError(t, err, "unexpected error while opening test exemplar storage")
-	return &TestStorage{DB: db, exemplarStorage: es, dir: dir}
+	if err != nil {
+		return nil, fmt.Errorf("opening test exemplar storage: %w", err)
+	}
+	return &TestStorage{DB: db, exemplarStorage: es, dir: dir}, nil
 }
 
 type TestStorage struct {

